IDR/notebook/scripts/nbconverted/2.collect_statistics.py

```diff
@@ -0,0 +1,354 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[15]:
+
+
+import pandas as pd
+import pathlib
+import numpy as np
+from numpy import log as ln
+import scipy.stats as stats
+from scipy import ndimage
+import multiprocessing
+import time
+
+
+# In[16]:
+
+
+def walk(path):
+    """Collects paths for files in path parameter
+
+    Parameters
+    ----------
+    path: str or pathlib.Path() object
+            Path to metadata folder containing IDR study directories
+
+    Returns
+    -------
+    PosixPath object
+    """
+    for subdir in pathlib.Path(path).iterdir():
+        if subdir.is_dir():
+            yield from walk(subdir)
+            continue
+        yield subdir.resolve()
+
+
+# In[17]:
+
+
+# Define stats
+def category_frequencies(attribute_elements):
+    """Calculates absolute and relative frequencies for unique elements of an image attribute
+
+    Parameters
+    ----------
+    attribute_elements: dict
+        Unique elements of an image attribute as keys and instances of each as values
+
+    Returns
+    -------
+    rel_freq_dict: dict
+        Relative frequencies per unique image attribute element
+    abs_freq-list: list
+        Instance counts of unique image attribute elements
+    """
+    total_instances = sum(attribute_elements.values())
+    rel_freq_dict = dict()
+    abs_freq_list = list()
+    for image_attribute in attribute_elements.keys():
+        abs_freq_list.append(attribute_elements[image_attribute])
+        rel_freq_dict[image_attribute] = (
+            attribute_elements[image_attribute] / total_instances
+        )
+    return rel_freq_dict, abs_freq_list
+
+
+def h_index(p):
+    """Calculates the Shannon Index of a set of unique attribute instances.
+
+    Parameters
+    ----------
+    p: dict
+        Dictionary of relative frequencies for each unique element in an attribute column.
+
+    Returns
+    -------
+    h: float
+        Shannon Index value
+    results: list
+        List of each -p_iln(p_i) value to use for Normalized Median Evenness statistic.
+    """
+    results = list()
+    for entry in p.values():
+        results.append(entry * ln(entry))
+
+    results = np.array(results)
+    h = -(sum(results))
```

Comment:
Where possible, I highly _highly_ recommend not reinventing the wheel. Is there a well used package that can calculate `h` for us? Maybe [`skbio`](http://scikit-bio.org/docs/0.4.2/generated/generated/skbio.diversity.alpha.shannon.html)?

We should need worry less about correct implementations or inaccurate assumptions if we use external packages with community backing.

I hope that this update doesn't cause much headache!

Reply:
@gwaybio So I was able to implement `skbio`'s versions of these calculations. However, they are still considered "experimental" and it shows with the gini coefficient as it outputs negative values, which isn't possible if done correctly. The gini coefficient function I used before came from the stackoverflow community and doesn't output negative values. 

`skbio` also requires a previous version of `scipy` because of import errors which the devs say won't be fixed until October probably.

Reply:
@gwaybio I've never had good experiences with `scikit` anything and I'm not sure how to proceed here given these functions are still in development (experimental). What do you recommend?

Reply:
ugh, sorry to have sent you down a rabbit hole. Hopefully it was worth the exploration however, and good to know that the implementations are experimental (and perhaps wrong!)

In these cases, you can proceed in the following way:

1. Revert back to your own implementation
2. Potentially contribute your implementations back into `skbio`

Reply:
@gwaybio Where contributing back would be forking from their repo and submitting a PR?

Reply:
Yep, exactly the same general procedures we have in the lab. (There are likely other specific requirements as well)

Also note that this is absolutely not required for you to do 🙂

Reply:
@gwaybio Nah that's real cool. I'll do that after this (maybe the next) PR.

Reply:
@gwaybio Stat's workflow finalized. I reverted back to what I was doing before and added an exception for if richness (s) `s == 1` because it causes division by 0 errors. 

If `s==1`, then `h = 0`, `nme = None`, and `j = None` since `h = -plnp` where `p = rel_freq = [1.0]`. 

For nme and j, they divide by `ln(s)`, causing division by 0.

Added in 0a1bed2608bec4717dc55527275cc12faab191ad.